Handle owncloud migration to latest release

juliusknorr opened the pull request:
Steps to migrate from 10.5 to this branch
- make sure to have activity/files_pdfviewer/notifications/viewer/firstrunwizard installed
- occ upgrade
- occ db:convert-filecache-bigint
- occ db:add-missing-columns
- occ db:add-missing-indices
- occ db:add-missing-primary-keys (edit by PVince81, had to do this on my setup)

## Adjustments

- [x] Avoid exception in bruteforce protection capabilities on the upgrade page (as the table doesn't exist yet)
- [x] db:add-missing-indices
  In SchemaException.php line 85: There is no column with name 'userid' on table 'oc_properties'.  
- [x] Check for removed migrations/repair in server steps between 12 and master
- [x] Revert repair steps that might be relevant when migrating 

## Missing from the database diff between 10.5 and 20

With this set of patches the migration succeed however there is some difference in the database schema that should be addressed:


- [x] oc_authtoken -> loginname -> 255, add scope field
- [x] oc_calendar components 20 -> 64
- [x] oc_calendarsubscriptions lastmodified should be default null
- [x] oc_cards -> add KEY `addressbookid_uri_index` (`addressbookid`,`uri`),
- [x] oc_external_config -> value default null
- [x] oc_files_trash auto_id bigint
- [x] oc_jobs execution duration default -1/0
- [x] oc_mounts -> add KEY `mounts_mount_id_index` (`mount_id`)
- [x] oc_privatedata (can be ingored, moved to https://github.com/nextcloud/privatedata)
- [x] oc_dav_properties -> migrate to oc_properties (see https://github.com/nextcloud/server/pull/23044#issuecomment-718807851) The table is kept for backup reasons
- [x] oc_properties -> droped file id and contents
- [x] oc_trusted_servers (only comments added

## Removing data from database (core/Migrations/Version21000Date20201120141228.php)

- [x] oc_dav_job_status -> delete this table, see https://github.com/nextcloud/server/pull/23044#issuecomment-718804467
- [x] oc_systemtags -> remove assignable
- [x] oc_share -> remove attributes




## Pulled out

- [x] oc_federated_reshares -> move remote_id to string (https://github.com/nextcloud/server/pull/24247)
- [x] oc_federated_reshares -> share_id to bigint (https://github.com/nextcloud/server/pull/24247)
- [ ] oc_activity -> app nc: varchar(32) oc: varchar(255) , add type to activity_filter key
- [ ] oc_notifications -> only index names are different

juliusknorr committed:
Add missing table columns

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Allow upgrade from oc 10.5

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Drop fk constraints on locks table

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Fix missing authtoken scope

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Avoid checking for brute force protection capabilities when upgrading

This might happen a releases that doesn't have this table yet

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Allow major/minor match for owncloud version

Signed-off-by: Julius Härtl <jus@bitgrid.net>

PVince81 commented:
> oc_activity -> app varchar 32 -> 255, add type to activity_filter key

In a fresh install of NC 20 the app column still has 32 characters, same in activity master.
Not sure where that comes from.

PVince81 commented:
ok, got it, it's the other way around...

PVince81 committed:
Adjust calendars.components to 64 chars

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Adjust further columns

- calendarsubscriptions.lastmodified -> not null
- external_config.value -> allow null

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
regarding bigints I think we should add these to NC (and not convert bigint to int in NC)
the reason is most likely that the number of fileids/entries can become very high on huge instances

not sure if such change would be backported to NC 20 (if DB changes are allowed in minor releases), so might only arrive with NC 21 then ?

PVince81 commented:
oc_dav_job_status is for pending asynchronous/lazy operations: https://github.com/owncloud/core/blob/master/apps/dav/lib/DAV/LazyOpsPlugin.php#L65

we can delete it as it shouldn't contain any active entries during a migration

PVince81 commented:
oc_dav_properties is using fileid to map DAV properties so that they can survive renames

given that oc_properties only contains the path, DAV properties will likely get lost when renaming through some code paths.
see https://github.com/owncloud/core/issues/25563

maybe to keep oc_properties we could put a file id into the "propertiespath" column ?

PVince81 committed:
Adjust execution duration to 0

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

juliushaertl commented:
Thanks a lot for digging into this already :) 

> regarding bigints I think we should add these to NC (and not convert bigint to int in NC)

Yes, definitely. At least bigint conversions are fine for minor releases if we let the user do the migration through `occ db:convert-filecache-bigint`

> not sure if such change would be backported to NC 20 (if DB changes are allowed in minor releases), so might only arrive with NC 21 then ?

@rullzer what do you think?

juliusknorr committed:
Drop oc_dav_job_status table

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Drop assignable systemtag column

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Drop attributes on oc_share table

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Add missing mount_id index

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Move authtoken login_name column to varchar(255)

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Add missing index on oc_cards and rename if it previously existed

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliushaertl commented:
> Or in theory we could also discard (or keep for later) that table, in case whatever properties exist in there is anyway irrelevant for NC apps. Note that sometimes when Webdav is mounted as a Windows drive it will add a lot of Windows-specific properties. Deleting these shouldn't cause any issue as Windows would re-populate the next time.

I'd actually say we can go that way. We just recreate the table then with our format and mainly need to migrate oc_dav_properties from that to ours. For that migration I'm not entirely sure if there is any other relevant properties than calendar ones (cc @skjnldsv does store contacts some in there as well) we could just extract the user id from the propertypath, but my DAV internal knowledge is a bit limited in that regard. So any input welcome. Maybe @icewind1991 or @rullzer can also have a look if that makes sense.


For the ownCloud oc_dav_properties migration to oc_properties i would just try to match against /calendar/userid/ and extract the userid for our table format from that.
```
+----+------------------------+------------------------------------------+---------------+--------------+
| id | propertypath           | propertyname                             | propertyvalue | propertytype |
+----+------------------------+------------------------------------------+---------------+--------------+
|  1 | calendars/admin/sdfsdf | {http://owncloud.org/ns}calendar-enabled | 0             |            1 |
+----+------------------------+------------------------------------------+---------------+--------------+

becomes

+----+--------+--------------------------+------------------------------------------+---------------+
| id | userid | propertypath             | propertyname                             | propertyvalue |
+----+--------+--------------------------+------------------------------------------+---------------+
|  1 | jus    | calendars/admin/personal | {http://owncloud.org/ns}calendar-enabled | 0             |
+----+--------+--------------------------+------------------------------------------+---------------+

```

The propertytype introduced in https://github.com/owncloud/core/pull/37314 is also most likely irrelevant for us for now.

rullzer commented:
How did you get to that userid? shouldn't it be admin then?

juliushaertl commented:
Woops, must have been a copy past error when trying to align copied dumps from the the two instances i had. Yes that would of course be admin.

juliusknorr committed:
Handle oc_dav_properties migration

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Readd repair steps that are relevant when migrating from ownCloud

This reverts commit d9b1492e03ab9fe58bb87baaeba745790ca15c53.

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Dump autoloader

Signed-off-by: Julius Härtl <jus@bitgrid.net>

nickvergessen commented on core/Migrations/Version13000Date20170718121200.php:
```diff
@@ -454,6 +518,12 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
+		} else {
+			$table = $schema->getTable('jobs');
+			$table->changeColumn('execution_duration', [
+				'notnull' => true,
```
:see_no_evil: should be 
```suggestion
				'notnull' => false,
```

Otherwise we can't save when the duration was <1 second

nickvergessen commented on core/Migrations/Version13000Date20170718121200.php:
```diff
@@ -454,6 +518,12 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
+			$table = $schema->getTable('jobs');
+			$table->changeColumn('execution_duration', [
+				'notnull' => true,
+				'default' => 0,
```
invalid default for oracle, as you said "not null"

nickvergessen commented on core/Migrations/Version13000Date20170919121250.php:
```diff
@@ -63,8 +63,17 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
+			$column->setUnsigned(true);
+		} else {
+			$table->addColumn('remember', 'smallint', [
+				'notnull' => true,
```
invalid default

juliushaertl replied:
Fixed, there were actually some more columns affected by that in the authtoken table

juliusknorr committed:
Make sure the migrations table schema is always checked

Signed-off-by: Julius Härtl <jus@bitgrid.net>

Revert "Make sure the migrations table schema is always checked"

This reverts commit 258955ef738a52d9da2ac2fe59466e6093d7e9bc.

Set current vendor during upgrade and perform migrations table change if needed

Signed-off-by: Julius Härtl <jus@bitgrid.net>

nickvergessen commented on lib/private/DB/MigrationService.php:
```diff
@@ -124,11 +124,6 @@ private function createMigrationTable() {
 			return false;
 		}
 
-		if ($this->connection->tableExists('migrations')) {
```
:/ + 20 seconds for each app on installation. 
What's the diff with OC?

juliushaertl replied:
See the code below, basically the version and app id size doesn't match, meaning if we don't increase the size of the versions column the migrations will fail. Any other idea how we could solve that?

nickvergessen replied:
maybe wrap with something along:

https://github.com/nextcloud/server/blob/89ed2c37bf656ceb772bb6759c8977a7dc78b3fb/apps/files_sharing/lib/Migration/OwncloudGuestShareType.php#L78-L80

nickvergessen replied:
arg, dont have appconfig I guess

juliushaertl replied:
Ok, but we can just ignore the check if config is not available (e.g. on first install). Let me have  a look.

juliushaertl replied:
https://github.com/nextcloud/server/pull/23044/commits/f6da1c3f7b528c55fb2ad06e2038b3bfa2523133 seems to do the trick there.

nickvergessen replied:
😻

juliusknorr committed:
Properly migrate from new owncloud avatar location

as the files are not scanned we cannot use the OCP\Files api

Signed-off-by: Julius Härtl <jus@bitgrid.net>

juliusknorr committed:
Change further columns to be nullable with a default of 0

Signed-off-by: Julius Härtl <jus@bitgrid.net>

nickvergessen commented on lib/private/Repair/Owncloud/MoveAvatarsBackgroundJob.php:
```diff
@@ -0,0 +1,117 @@
+ */
+namespace OC\Repair\Owncloud;
+
+use OC\BackgroundJob\QueuedJob;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotFoundException;
+use OCP\Files\Storage;
+use OCP\IAvatarManager;
+use OCP\ILogger;
+use OCP\IUser;
+use OCP\IUserManager;
+use Psr\Log\LoggerInterface;
```
something is unused

nickvergessen reviewed (approved):
🐘

rullzer commented:
/backport to stable20

PVince81 commented:
seems we missed oc_share_external: https://github.com/nextcloud/server/issues/24813

PVince81 commented on core/Migrations/Version21000Date20201120141228.php:
```diff
@@ -0,0 +1,65 @@
+
+		if ($schema->hasTable('systemtag')) {
+			$table = $schema->getTable('systemtag');
+			if ($table->hasColumn('systemtag')) {
```
wrong check, this should be "hasColumn('assignable')"

dossantg commented:
Is it possible to backport this PR for 19.0.7? I think the commit (c77e259) for indice on oc_cards can fix https://github.com/nextcloud/server/issues/23980 

But maybe i am wrong!
